pallets/ajuna-affiliate/src/lib.rs

```diff
@@ -0,0 +1,269 @@
+		CannotAffiliateMoreAccounts,
+		/// This account has already been affiliated by another affiliator
+		CannotAffiliateAlreadyAffiliatedAccount,
+		/// This account is already an affiliator, so it cannot affiliate to another account
+		CannotAffiliateExistingAffiliator,
+		/// The account is blocked, so it cannot be affiliated to
+		CannotAffiliateBlocked,
+		/// The given extrinsic identifier is already paired with an affiliate rule
+		ExtrinsicAlreadyHasRule,
+	}
+
+	impl<T: Config<I>, I: 'static> Pallet<T, I> {
+		fn add_new_affiliate_to(
+			affiliator: T::AccountId,
+			affiliatee: T::AccountId,
+		) -> DispatchResult {
+			let accounts = {
+				let mut accounts_vec = Affiliatees::<T, I>::take(&affiliator).unwrap_or_default();
+
+				Self::try_add_account_to(&mut accounts_vec, affiliator.clone())?;
+
+				accounts_vec
+			};
+
+			Affiliatees::<T, I>::insert(affiliatee, accounts);
+			Affiliators::<T, I>::try_mutate(&affiliator, |state| {
+				state.affiliates = state
+					.affiliates
+					.checked_add(1)
+					.ok_or(DispatchError::Arithmetic(ArithmeticError::Overflow))?;
+
+				Ok(())
+			})
+		}
+
+		fn try_add_account_to(
+			accounts: &mut AffiliatedAccountsOf<T, I>,
+			account: T::AccountId,
+		) -> DispatchResult {
+			if accounts.len() == T::AffiliateMaxLevel::get() as usize {
+				accounts.pop();
+			}
+			accounts
+				.try_insert(0, account)
+				.map_err(|_| Error::<T, I>::CannotAffiliateMoreAccounts.into())
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> AffiliateInspector<AccountIdOf<T>> for Pallet<T, I> {
+		fn get_affiliator_chain_for(account: &AccountIdOf<T>) -> Option<Vec<AccountIdOf<T>>> {
+			Affiliatees::<T, I>::get(account).map(|accounts| accounts.into_inner())
+		}
+
+		fn get_affiliate_count_for(account: &AccountIdOf<T>) -> u32 {
+			Affiliators::<T, I>::get(account).affiliates
+		}
+	}
+
+	impl<T: Config<I>, I: 'static> AffiliateMutator<AccountIdOf<T>> for Pallet<T, I> {
+		fn try_mark_account_as_affiliatable(account: &AccountIdOf<T>) -> DispatchResult {
+			Affiliators::<T, I>::try_mutate(account, |state| {
+				ensure!(
+					state.status != AffiliatableStatus::Blocked,
+					Error::<T, I>::CannotAffiliateBlocked
+				);
+
+				state.status = AffiliatableStatus::Affiliatable;
+
+				Ok(())
+			})
+		}
+
+		fn force_mark_account_as_affiliatable(account: &AccountIdOf<T>) {
+			Affiliators::<T, I>::mutate(account, |state| {
+				state.status = AffiliatableStatus::Affiliatable;
+			});
+		}
+
+		fn mark_account_as_blocked(account: &AccountIdOf<T>) {
+			Affiliators::<T, I>::mutate(account, |state| {
+				state.status = AffiliatableStatus::Blocked;
+			});
+		}
+
+		fn try_add_affiliate_to(
+			account: &AccountIdOf<T>,
+			affiliate: &AccountIdOf<T>,
+		) -> DispatchResult {
+			ensure!(account != affiliate, Error::<T, I>::CannotAffiliateSelf);
+
+			let affiliate_state = Affiliators::<T, I>::get(affiliate);
+			ensure!(
+				affiliate_state.affiliates == 0,
+				Error::<T, I>::CannotAffiliateExistingAffiliator
```

Comment:
This error is a bit too cryptic for me without preposition. So I would add either `AffiliateWith` or `AffiliateTo` existing affiliator.